### PR TITLE
Enabling faster latency for validator execution

### PR DESCRIPTION
**Pull request by ShreyaR**

### CODE REVIEW

File: validator/main.py [hunk 1]
```diff
@@ -59,36 +60,45 @@ class RestrictToTopic(Validator):
...
         disable_llm (bool, Optional, defaults to False): controls whether to use
             the LLM fallback. At least one of disable_classifier and
             disable_llm must be False.
-        model_threshold (float, Optional, defaults to 0.5): The threshold used to
+        zero_shot_threshold (float, Optional, defaults to 0.5): The threshold used to
             determine whether to accept a topic from the Zero-Shot model. Must be
             a number between 0 and 1.
+        llm_threshold (int, Optional, defaults to 3): The threshold used to determine
+        if a topic exists based on the provided llm api. Must be between 0 and 5.
     """
```
Comment (ovejabu) [inline validator/main.py:74]: I think Args should be updated now that device not only takes int

**Review by ovejabu — APPROVED** [review]
Great job!! A description of the PR to have a summary of the improvements and the quantification of the latency improvement would be nice.

File: validator/main.py (hunk 1 above)
Comment (wylansford) [reply validator/main.py:74]: device is a Optional[Union[str, int]] now to support cpu/mps or an int for device

**Comment by wylansford** [comment]
This update makes some major latency improvements through a few different upgrades.

1. Improves overall throughput by converting the model to batch mode on a per topic basis. Instead of multiple calls for each topic, all categories are check in one pass.

2. The LLM call is now defaulting to gpt4o.

3. The LLM call has a shorter prompt that just returns a json list, improving latency significantly. 
note: We tried using function calling to improve the reliability of the llm, but it is much slower. Using json mode + telling the model to use json is faster than relying on function calling.

4. Overall code cleanup for readability and maintainability 

Most recent tests show ~700ms for the llm call, and around 1.5s total for inference on my m2 based mac on cpu. With a gpu it should be faster.

disabling the llm and with a gpu, latency can be as low as 300ms for a single validation call

File: validator/main.py (hunk 1 above)
Comment (ovejabu) [reply validator/main.py:74]: Exactly. What I meant is here in the docstrings it still appear as 
```
device (int, Optional, defaults to -1): Device ordinal for CPU/GPU
supports for Zero-Shot classifier. Setting this to -1 will leverage
CPU, a positive will run the Zero-Shot model on the associated CUDA
device id.
```
Also the default for llm_callable needs to be updated as it is now gpt4o

**Review by ovejabu — APPROVED** [review]

File: validator/main.py [hunk 2]
```diff
@@ -59,36 +60,45 @@ class RestrictToTopic(Validator):
...
+        device: Optional[Union[str, int]] = -1,
         model: Optional[str] = "facebook/bart-large-mnli",
         llm_callable: Union[str, Callable, None] = None,
         disable_classifier: Optional[bool] = False,
+        classifier_api_endpoint: Optional[str] = None,
         disable_llm: Optional[bool] = False,
         on_fail: Optional[Callable[..., Any]] = None,
-        model_threshold: Optional[float] = 0.5,
+        zero_shot_threshold: Optional[float] = 0.5,
+        llm_theshold: Optional[int] = 3,
```
Comment (ovejabu) [inline validator/main.py]: I believe there is a typo here and it should say llm_threshold

File: validator/main.py [hunk 3]
```diff
@@ -97,47 +107,101 @@ def __init__(
...
+            self._classifier = pipeline(
+                "zero-shot-classification",
+                model=self._model,
+                device=self._device,
+                hypothesis_template="This example has to do with topic {}.",
+                multi_label=True,
+            )
         else:
-            return self.get_topic_llm(text, candidate_topics)
+            # TODO api endpoint
```
Comment (ovejabu) [inline validator/main.py:138]: Will you address this TODO in another PR?

File: validator/main.py (hunk 1 above)
Comment (wylansford) [reply validator/main.py:74]: Oh yeah good catch. Sorry, I missed that. Fixed

File: validator/main.py (hunk 2 above)
Comment (wylansford) [reply validator/main.py]: fixed!

File: validator/main.py (hunk 3 above)
Comment (wylansford) [reply validator/main.py:138]: Yes, there is actually a PR out for this for a basic implementation. However we are scoping out a more permanent solution that will hopefully make this cleaner. Still a WIP

**Pull request merged**